### PR TITLE
WAM integration

### PR DESCRIPTION
**Pull request by bgavrilMS**
- [x] Integration with current broker abstraction on .net framework
- [x] Account picker on Interactive when account / login hint is not provided
- [x] AAD support
- [x] MSA support (without GetAccounts)
- [x] Demo WinForms app

- [x] Switch_account flow (i.e. when a login hint is used but user chooses a different account - currently an exception is created, but it should be treated as success)
- [x] MSA account ID write-back 
- [ ] MSA passthrough
- [x] .net core support
- [x] UWP support
- [ ] IWA
- [ ] Console app support (currently AcquireTokenInteractive with account picker is not working)
- [x] Spec and docs
- [x] Testing

Backlog

POP
ROPC

### CODE REVIEW

**Comment by bgavrilMS** [comment]
Account picker:

account_picker

File: src/client/Microsoft.Identity.Client/ClientApplicationBase.cs [hunk 1]
```diff
@@ -174,18 +174,49 @@ private async Task<IEnumerable<IAccount>> GetAccountsInternalAsync(ApiIds apiId,
...
+                brokerAccounts = await FilterBrokerAccountsByEnvAsync(brokerAccounts).ConfigureAwait(false);
+                return brokerAccounts;
+            }
+
+            return Enumerable.Empty<IAccount>();
+        }
+
+        private async Task<IEnumerable<IAccount>> FilterBrokerAccountsByEnvAsync(IEnumerable<IAccount> brokerAccounts)
+        {
+            ServiceBundle.DefaultLogger.Verbose($"Filtering broker accounts by env. Before fitering: " + brokerAccounts.Count());
```
Comment (jennyf19) [inline src/client/Microsoft.Identity.Client/ClientApplicationBase.cs:186]: null check for brokerAccounts

Comment (jennyf19) [inline src/client/Microsoft.Identity.Client/ClientApplicationBase.cs:186]: >fitering [](start = 91, length = 8)

sp: filtering :)

File: src/client/Microsoft.Identity.Client/ClientApplicationBase.cs [hunk 2]
```diff
@@ -174,18 +174,49 @@ private async Task<IEnumerable<IAccount>> GetAccountsInternalAsync(ApiIds apiId,
...
+
+            ISet<string> allEnvs = new HashSet<string>(
+                brokerAccounts.Select(aci => aci.Environment),
+                StringComparer.OrdinalIgnoreCase);
+
+            var instanceMetadata = await ServiceBundle.InstanceDiscoveryManager.GetMetadataEntryTryAvoidNetworkAsync(
+                Authority,
+                allEnvs,
+                CreateRequestContext(Guid.NewGuid())).ConfigureAwait(false);
+
```
Comment (jennyf19) [inline src/client/Microsoft.Identity.Client/ClientApplicationBase.cs:196]: nit: is there a reason to create a new guid here and not use the RC from the service bundle?

File: src/client/Microsoft.Identity.Client/ClientApplicationBase.cs [hunk 3]
```diff
@@ -174,18 +174,49 @@ private async Task<IEnumerable<IAccount>> GetAccountsInternalAsync(ApiIds apiId,
...
+
+            brokerAccounts = brokerAccounts.Where(acc => instanceMetadata.Aliases.ContainsOrdinalIgnoreCase(acc.Environment));
+
+            ServiceBundle.DefaultLogger.Verbose($"After filtering: " + brokerAccounts.Count());
```
Comment (jennyf19) [inline src/client/Microsoft.Identity.Client/ClientApplicationBase.cs:200]: null check here

**Comment by jennyf19** [comment]
> // Copyright (c) Microsoft Corporation. All rights reserved.

nit: maybe a comment here, copied from the xml from ClientApplicationBase

---
Refers to: src/client/Microsoft.Identity.Client/IAccountInternal.cs:1 in e02a35c. [](commit_id = e02a35cd11d2e5a51bdcde396b5a2400e479856f, deletion_comment = False)

File: src/client/Microsoft.Identity.Client/IAccountInternal.cs [hunk 4]
```diff
@@ -0,0 +1,12 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System.Collections.Generic;
+
+namespace Microsoft.Identity.Client
+{
+    internal interface IAccountInternal : IAccount
+    {
+        IDictionary<string, string> WamAccountIds { get; }
```
Comment (jennyf19) [inline src/client/Microsoft.Identity.Client/IAccountInternal.cs:10]: nit: maybe a comment here, copied from the xml from ClientApplicationBase

File: src/client/Microsoft.Identity.Client/MsalError.cs [hunk 5]
```diff
@@ -908,5 +908,15 @@ public static class MsalError
...
+
+        /// <summary>
+        /// When calling AcquireTokenInteractive with the WAM broker, the call must be made from the UI thread.
+        /// </summary>
+        public const string WamUiThread = "wam_ui_thread_only";
+
+        /// <summary>
+        /// The Windows broker (WAM) is not supported in conjunction with B2C authorities
+        /// </summary>
+        public const string WamNoB2C = "wam_no_b2c";
```
Comment (jennyf19) [inline src/client/Microsoft.Identity.Client/MsalError.cs:920]: >WamNoB2C [](start = 28, length = 8)

nit: we're trying to move away from "b2c" branding in the api...is there a different way to say it w/out b2c? `WamNotSupported` or something? Assume you'll have an aka.ms link somewhere w/more details.

File: src/client/Microsoft.Identity.Client/ApiConfig/AcquireTokenInteractiveParameterBuilder.cs [hunk 6]
```diff
@@ -338,7 +338,7 @@ public AcquireTokenInteractiveParameterBuilder WithParentActivityOrWindow(NSWind
         protected override void Validate()
         {
             base.Validate();
-
+            // this.PublicClientApplicationExecutor.ServiceBundle.Config.IsBrokerEnabled
```
Comment (jennyf19) [inline src/client/Microsoft.Identity.Client/ApiConfig/AcquireTokenInteractiveParameterBuilder.cs:341]: delete?

File: src/client/Microsoft.Identity.Client/ApiConfig/Parameters/AcquireTokenInteractiveParameters.cs [hunk 7]
```diff
@@ -13,7 +13,7 @@ namespace Microsoft.Identity.Client.ApiConfig.Parameters
 
     internal class AcquireTokenInteractiveParameters : IAcquireTokenParameters
     {
-        public Prompt Prompt { get; set; } = Prompt.SelectAccount;
+        public Prompt Prompt { get; set; } = Prompt.NotSpecified;
```
Comment (jennyf19) [inline src/client/Microsoft.Identity.Client/ApiConfig/Parameters/AcquireTokenInteractiveParameters.cs:16]: this is interesting. is this a breaking change?

File: src/client/Microsoft.Identity.Client/Internal/Requests/AuthenticationRequestParameters.cs [hunk 8]
```diff
@@ -68,9 +70,25 @@ internal class AuthenticationRequestParameters
         public ApiEvent.ApiIds ApiId => _commonParameters.ApiId;
 
         public RequestContext RequestContext { get; }
+
+        /// <summary>
+        /// Authority used by MSAL for most operations. After the /token call, the tenant ID is 
+        /// always known and TenantUpdatedCanonicalAuthority is created and used.
+        /// In Silent flows, the Authority cannot be unkwon ("common" or "organizations") and
```
Comment (jennyf19) [inline src/client/Microsoft.Identity.Client/Internal/Requests/AuthenticationRequestParameters.cs:77]: >nkwon [](start = 54, length = 5)

sp: unkown

File: src/client/Microsoft.Identity.Client/Internal/Requests/Interactive/AuthCodeRequestComponent.cs [hunk 9]
```diff
@@ -176,7 +176,11 @@ public Uri GetAuthorizationUriWithoutPkce()
                 authorizationRequestParameters[kvp.Key] = kvp.Value;
             }
 
-            if (_interactiveParameters.Prompt.PromptValue != Prompt.NoPrompt.PromptValue)
+            if (_interactiveParameters.Prompt == Prompt.NotSpecified)
```
Comment (jennyf19) [inline src/client/Microsoft.Identity.Client/Internal/Requests/Interactive/AuthCodeRequestComponent.cs:179]: why have the NotSpecified if it's using selectAccount? i'd think if developers did not set a prompt we shouldn't send any prompt value (that's actually what b2c would prefer at least, so they (ests) can determine which sign-in screen to show based on the current session cookie). I don't see the value in adding a new prompt value.

File: src/client/Microsoft.Identity.Client/Platforms/Features/WamBroker/AadPlugin.cs [hunk 10]
```diff
@@ -0,0 +1,228 @@
...
+        public AadPlugin(IWamProxy wamProxy, IWebAccountProviderFactory webAccountProviderFactory, ICoreLogger logger)
+        {
+            _wamProxy = wamProxy;
+            _webAccountProviderFactory = webAccountProviderFactory;
+            _logger = logger;
+        }
+
+        public async Task<IEnumerable<IAccount>> GetAccountsAsync(string clientID)
+        {
+            var webAccounProvider = await _webAccountProviderFactory.GetAccountProviderAsync("organizations").ConfigureAwait(false);
```
Comment (jennyf19) [inline src/client/Microsoft.Identity.Client/Platforms/Features/WamBroker/AadPlugin.cs:32]: >"organizations" [](start = 93, length = 15)

nit: you have a constant value for this

Comment (jennyf19) [inline src/client/Microsoft.Identity.Client/Platforms/Features/WamBroker/AadPlugin.cs:32]: >webAccounProvider [](start = 16, length = 17)

sp: webAccountProvider

File: src/client/Microsoft.Identity.Client/Platforms/Features/WamBroker/AadPlugin.cs [hunk 11]
```diff
@@ -0,0 +1,228 @@
...
+
+            var webAccounts = await _wamProxy.FindAllWebAccountsAsync(webAccounProvider, clientID).ConfigureAwait(false);
+
+            var msalAccounts = webAccounts
+                .Select(webAcc => ConvertToMsalAccountOrNull(webAcc))
+                .Where(a => a != null)
+                .ToList();
+
+            _logger.Info($"[WAM AAD Provider] GetAccountsAsync converted {webAccounts.Count()} MSAL accounts");
```
Comment (jennyf19) [inline src/client/Microsoft.Identity.Client/Platforms/Features/WamBroker/AadPlugin.cs:41]: >webAccounts [](start = 74, length = 11)

null check here?

File: src/client/Microsoft.Identity.Client/Platforms/Features/WamBroker/AadPlugin.cs [hunk 12]
```diff
@@ -0,0 +1,228 @@
...
+            return msalAccounts;
+        }
+
+
+        private Account ConvertToMsalAccountOrNull(WebAccount webAccount)
+        {
+            string username = webAccount.UserName;
```
Comment (jennyf19) [inline src/client/Microsoft.Identity.Client/Platforms/Features/WamBroker/AadPlugin.cs:48]: null check on webAccount?

**Comment by jennyf19** [comment]
i'm about part-way through. looks awesome so far. i'll pull your branch and try it.

---
In reply to: [701247755](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/1962#issuecomment-701247755) [](ancestors = 701247755)

File: src/client/Microsoft.Identity.Client/ClientApplicationBase.cs (hunk 1 above)
Comment (bgavrilMS) [reply src/client/Microsoft.Identity.Client/ClientApplicationBase.cs:186]: Fixed.

File: src/client/Microsoft.Identity.Client/MsalError.cs (hunk 5 above)
Comment (bgavrilMS) [reply src/client/Microsoft.Identity.Client/MsalError.cs:920]: ack

Comment (bgavrilMS) [reply src/client/Microsoft.Identity.Client/MsalError.cs:920]: I will use the Windows terminology, smth like "The Windows broker (WAM) is only supported in conjunction with "work and school" accounts and with Microsoft accounts."

File: src/client/Microsoft.Identity.Client/ApiConfig/Parameters/AcquireTokenInteractiveParameters.cs (hunk 7 above)
Comment (bgavrilMS) [reply src/client/Microsoft.Identity.Client/ApiConfig/Parameters/AcquireTokenInteractiveParameters.cs:16]: Nice catch 👍 
This is not a breaking change, as "NotSpecified" is translated to "SelectAccount" when broker is not in use. I've noticed that we don't really offer the users the possibility of not sending a promp option (except via NoPrompt which is documented as for B2C use only), so maybe we will want to take this a breaking change. I logged https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/issues/2086 to track this.

File: src/client/Microsoft.Identity.Client/Platforms/Features/WamBroker/AadPlugin.cs (hunk 10 above)
Comment (bgavrilMS) [reply src/client/Microsoft.Identity.Client/Platforms/Features/WamBroker/AadPlugin.cs:32]: There are too many important constants in this implementation for them to be hidden away in some file. I'm gonna let them shine in the code :)

**Comment by bgavrilMS** [comment]
Thank you Jenny for the review so far. I think this is too much code to review in depth, so I did some presentations instead. I'm happy to repeat these for you if you wish.

File: src/client/Microsoft.Identity.Client/MsalError.cs (hunk 5 above)
Comment (jennyf19) [reply src/client/Microsoft.Identity.Client/MsalError.cs:927]: good idea.

**Comment by jennyf19** [comment]
@bgavrilMS yes! i would definitely like to have the presentations as well. didn't realize i had missed them. thanks! i'll set something up.

**Pull request merged**